### PR TITLE
refactor: pass context for goroutine shutdown

### PR DESCRIPTION
**Pull request by 0xERR0R**
- pass context to server/resolvers
- each goroutine checks for cancellation
- I think, maybe we'll need it for config reload -> we can just close the context and recreate all resolvers

### CODE REVIEW

File: cmd/serve.go [hunk 1]
```diff
@@ -43,15 +44,19 @@ func startServer(_ *cobra.Command, _ []string) error {
 
 	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
 
-	srv, err := server.NewServer(cfg)
+	ctx, cancelFn := context.WithCancel(context.Background())
```
Comment (ThinkChaos) [inline cmd/serve.go:47]: We can `defer` the `cancelFn` here to make this more foolproof :)

File: config/config.go [hunk 2]
```diff
@@ -331,20 +334,29 @@ func (c *SourceLoadingConfig) StartPeriodicRefresh(refresh func(context.Context)
...
 	defer ticker.Stop()
 
-	for range ticker.C {
-		err := refresh(context.Background())
-		if err != nil {
-			logErr(err)
+	for {
+		select {
+		case <-ticker.C:
+			err := refresh(context.Background())
```
Comment (ThinkChaos) [inline config/config.go:353]:
```suggestion
			err := refresh(ctx)
```

**Review by ThinkChaos — COMMENTED** [review]
This is great!

**Comment by kwitsch** [comment]
Agreed this looks really great. 👍

File: cmd/serve.go (hunk 1 above)
Comment (0xERR0R) [reply cmd/serve.go:47]: yes, it makes absolutely sense. And we save one line of code :)

**Review by ThinkChaos — APPROVED** [review]

**Review by kwitsch — APPROVED** [review]

**Pull request merged**